File: z-clients/java/src/main/java/com/svix/diom/apis/Idempotency.java
// this file is @generated
package com.svix.diom.apis;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.svix.diom.DiomException;
import com.svix.diom.HttpClient;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import com.svix.diom.models.IdempotencyAbortIn;
import com.svix.diom.models.IdempotencyAbortOut;
import com.svix.diom.models.IdempotencyCompleteIn;
import com.svix.diom.models.IdempotencyCompleteOut;
import com.svix.diom.models.IdempotencyStartIn;
import com.svix.diom.models.IdempotencyStartOut;
import com.svix.diom.models.IdempotencyStartIn_;
import com.svix.diom.models.IdempotencyCompleteIn_;
import com.svix.diom.models.IdempotencyAbortIn_;

public class Idempotency {
    private final HttpClient client;

    public Idempotency(HttpClient client) {
        this.client = client;
    }

    public IdempotencyNamespace namespace() {
        return new IdempotencyNamespace(this.client);
    }

    /** Start an idempotent request */
    public IdempotencyStartOut start(
        String key,
        final IdempotencyStartIn idempotencyStartIn
    ) throws DiomException {
        IdempotencyStartIn_ body = new IdempotencyStartIn_(
            idempotencyStartIn.getNamespace(),
            key,
            idempotencyStartIn.getLockPeriod()
        );

        return this.client.executeRequest(
            "POST",
            "/api/v1.idempotency.start",
            null,
            body,
            IdempotencyStartOut.class
        );
    }

    /** Complete an idempotent request with a response */
    public IdempotencyCompleteOut complete(
        String key,
        final IdempotencyCompleteIn idempotencyCompleteIn
    ) throws DiomException {
        IdempotencyCompleteIn_ body = new IdempotencyCompleteIn_(
            idempotencyCompleteIn.getNamespace(),
            key,
            idempotencyCompleteIn.getResponse(),
            idempotencyCompleteIn.getContext(),
            idempotencyCompleteIn.getTtl()
        );

        return this.client.executeRequest(
            "POST",
            "/api/v1.idempotency.complete",
            null,
            body,
            IdempotencyCompleteOut.class
        );
    }

    /** Abandon an idempotent request (remove lock without saving response) */
    public IdempotencyAbortOut abort(
        String key,
        final IdempotencyAbortIn idempotencyAbortIn
    ) throws DiomException {
        IdempotencyAbortIn_ body = new IdempotencyAbortIn_(
            idempotencyAbortIn.getNamespace(),
            key
        );

        return this.client.executeRequest(
            "POST",
            "/api/v1.idempotency.abort",
            null,
            body,
            IdempotencyAbortOut.class
        );
    }

    /** Abandon an idempotent request (remove lock without saving response) */
    public IdempotencyAbortOut abort(
        String key
    ) throws DiomException {
        return this.abort(
            key,
            new IdempotencyAbortIn()
        );
    }
}

File: z-clients/javascript/src/apis/idempotency.ts
// this file is @generated

import {
    type IdempotencyAbortIn,
    IdempotencyAbortInSerializer,
} from '../models/idempotencyAbortIn';
import {
    type IdempotencyAbortOut,
    IdempotencyAbortOutSerializer,
} from '../models/idempotencyAbortOut';
import {
    type IdempotencyCompleteIn,
    IdempotencyCompleteInSerializer,
} from '../models/idempotencyCompleteIn';
import {
    type IdempotencyCompleteOut,
    IdempotencyCompleteOutSerializer,
} from '../models/idempotencyCompleteOut';
import {
    type IdempotencyStartIn,
    IdempotencyStartInSerializer,
} from '../models/idempotencyStartIn';
import {
    type IdempotencyStartOut,
    IdempotencyStartOutSerializer,
} from '../models/idempotencyStartOut';
import { IdempotencyNamespace } from './idempotencyNamespace';
import { HttpMethod, DiomRequest, type DiomRequestContext } from "../request";

export class Idempotency {
    public constructor(private readonly requestCtx: DiomRequestContext) {}

    public get namespace() {
        return new IdempotencyNamespace(this.requestCtx);
    }

    /** Start an idempotent request */
    public start(
        key: string,
        idempotencyStartIn: IdempotencyStartIn,
    ): Promise<IdempotencyStartOut> {
        const request = new DiomRequest(HttpMethod.POST, "/api/v1.idempotency.start");

        request.setBody(
            IdempotencyStartInSerializer._toJsonObject({
                ...idempotencyStartIn,
                key: key,
            })
        );
        
        return request.send(
            this.requestCtx,
            IdempotencyStartOutSerializer._fromJsonObject,
        );
    }/** Complete an idempotent request with a response */
    public complete(
        key: string,
        idempotencyCompleteIn: IdempotencyCompleteIn,
    ): Promise<IdempotencyCompleteOut> {
        const request = new DiomRequest(HttpMethod.POST, "/api/v1.idempotency.complete");

        request.setBody(
            IdempotencyCompleteInSerializer._toJsonObject({
                ...idempotencyCompleteIn,
                key: key,
            })
        );
        
        return request.send(
            this.requestCtx,
            IdempotencyCompleteOutSerializer._fromJsonObject,
        );
    }/** Abandon an idempotent request (remove lock without saving response) */
    public abort(
        key: string,
        idempotencyAbortIn: IdempotencyAbortIn,
    ): Promise<IdempotencyAbortOut> {
        const request = new DiomRequest(HttpMethod.POST, "/api/v1.idempotency.abort");

        request.setBody(
            IdempotencyAbortInSerializer._toJsonObject({
                ...idempotencyAbortIn,
                key: key,
            })
        );
        
        return request.send(
            this.requestCtx,
            IdempotencyAbortOutSerializer._fromJsonObject,
        );
    }
}

